Particles, Projectiles, and Extras: QOL Changes

lzardy opened the pull request:
1. Fixed not being able to have multiple projectiles per spell
2. Projectiles' OriginSpell and SpellData references are now public (read only)
3. Spells now contain references to all of their projectiles
4. Added particle visibility checks
5. Replaces projectile and particle packets with LeaguePackets' versions (implementations are likely incomplete, check comments for details)
6. Adds extra packets (just so it doesn't have to be done later)
7. Removes many of the unused PacketDefinitions420 packets (there are more, but decided to only remove ones that were replaced in this PR)
Close #831 and #308

lzardy committed:
Particles, Projectiles, and Extras: QOL Changes

1. Fixed not being able to have multiple projectiles per spell
2. Projectiles' OriginSpell and SpellData references are now public (read only)
3. Spells now contain references to all of their projectiles
4. Added particle visibility checks
5. Replaces projectile and particle packets with LeaguePackets' versions (implementations are likely incomplete, check comments for details)
6. Adds extra packets (just so it doesn't have to be done later)
7. Removes many of the unused PacketDefinitions420 packets (there are more, but decided to only remove ones that were replaced in this PR)

danil179 commented on GameServerCore/Domain/GameObjects/IProjectile.cs:
```diff
@@ -7,5 +7,8 @@ public interface IProjectile : IObjMissile
         IAttackableUnit Owner { get; }
         int ProjectileId { get; }
+        ISpellData SpellData { get; }
+        ISpell OriginSpell { get; }
```
So `ISpellData` isn't in `ISpell`? And why do we really want to expose it in the interface?

lzardy replied:
We can probably remove it from Projectile altogether since it has OriginSpell.

lzardy replied:
Nevermind, it is the projectile spell's spelldata, not the spell that casts it.

lzardy replied:
If it is part of the interface, scripts can use it for values, such as scaling, damage, particles, flags, etc. It would probably help to eliminate a lot of constants in scripts.

danil179 replied:
Seems ok.

danil179 commented on GameServerLib/GameObjects/Spells/Spell.cs:
```diff
@@ -95,7 +96,8 @@ public virtual bool Cast(float x, float y, float x2, float y2, IAttackableUnit u
             Y2 = y2;
             Target = u;
-            FutureProjNetId = _networkIdManager.GetNewNetId();
+            uint FutureProjNetId = _networkIdManager.GetNewNetId();
```
Not used?

lzardy replied:
It is used only in Spell.cs for NotifyCastSpell.

danil179 replied:
This is local variable

lzardy replied:
Yes, what about it?

danil179 replied:
Where do you use it?

danil179 replied:
Ok, I missed the usage.

danil179 replied:
Do you really not need it? (to insert it into the projectiles list)

lzardy replied:
If you want, I could delete the variable and put `_networkIdManager.GetNewNetId()` into `NotifyCastSpell`.

danil179 replied:
From what I remember, you need to add this id to the first projectile (like Nidalee spear). Check if Nidalee spear work with this PR.

lzardy replied:
Made it private and re-assigned netId after each projectile is added.

danil179 commented on GameServerLib/GameObjects/Spells/Spell.cs:
```diff
@@ -30,7 +31,7 @@ public class Spell : ISpell
         public float CurrentCastTime { get; protected set; }
         public float CurrentChannelDuration { get; protected set; }
-        public uint FutureProjNetId { get; protected set; }
+        public List<IProjectile> Projectiles { get; protected set; }
```
Consider different implementation like hash table.

lzardy replied:
So, a dictionary?

lzardy replied:
Should I make the key an int for their position in the dictionary or uint for the projectile's netID?

danil179 replied:
Yeah, by NetId.

danil179 commented on GameServerLib/GameObjects/Spells/Spell.cs:
```diff
@@ -231,17 +238,23 @@ public void AddProjectile(string nameMissile, float fromX, float fromY, float to
+                projNetId,
+                isServerOnly
             );
+            if (p != null)
```
Why `p` will be `null`?

danil179 replied:
And this is a good point to use `?.`

lzardy replied:
Only did it out of instinct, didn't have any cases yet where it was null.

lzardy replied:
Removed the check.

danil179 commented on GameServerLib/ObjectManager.cs:
```diff
@@ -59,6 +59,42 @@ public void Update(float diff)
+                var particle = obj as IParticle;
+                foreach (var team in Teams)
+                {
+                    if (particle != null)
```
Should be placed before the `foreach`

lzardy replied:
Done.

danil179 reviewed (commented):
I did fast review. I'll approve later.

lzardy committed:
Cleaning up

In ObjectManager.cs:
	Changed check for a null particle to be before doing `foreach team`.
In Spell.cs/ISpell.cs:
	1. Projectiles variable is now a dictionary housing each projectile added by the OriginSpell and each projectile's netId (netId, IProjectile).
	2. FutureProjNetId is now used for all netIds assigned to added projectiles.

danil179 commented on GameServerLib/GameObjects/Spells/Spell.cs:
```diff
@@ -43,6 +44,8 @@ public class Spell : ISpell
         private Game _game;
         protected NetworkIdManager _networkIdManager;
 
+        private uint FutureProjNetId;
```
Disagree with style guide.

danil179 reviewed (changes requested):
Code style problems.

lzardy committed:
Conform to Style Guidlines

In Spell.cs:
	Changes FutureProjNetId to _futureProjNetId.

danil179 reviewed (approved):
Approved.
